src/Microsoft.AspNet.OData.Shared/Formatter/Serialization/SelectExpandNode.cs

```diff
@@ -129,418 +202,471 @@ private void Initialize(SelectExpandClause selectExpandClause, IEdmStructuredTyp
             }
             else
             {
-                // So far, it includes all properties of complex and collection of complex
-                HashSet<IEdmStructuralProperty> allComplexStructuralProperties = new HashSet<IEdmStructuralProperty>();
-                GetStructuralProperties(structuredType, allStructuralProperties, allComplexStructuralProperties);
-
-                // So far, it includes all navigation properties
-                HashSet<IEdmNavigationProperty> allNavigationProperties;
-                HashSet<IEdmAction> allActions;
-                HashSet<IEdmFunction> allFunctions;
-                IEnumerable<SelectItem> selectItems = new List<SelectItem>();
-
-                if (entityType != null)
-                {
-                    allNavigationProperties = new HashSet<IEdmNavigationProperty>(entityType.NavigationProperties());
-                    allActions = new HashSet<IEdmAction>(model.GetAvailableActions(entityType));
-                    allFunctions = new HashSet<IEdmFunction>(model.GetAvailableFunctions(entityType));
-                }
-                else if (structuredType != null)
-                {
-                    allNavigationProperties = new HashSet<IEdmNavigationProperty>(structuredType.NavigationProperties());
-                    
-                    // Currently, the library does not support for bounded operations on complex type. 
-                    allActions = new HashSet<IEdmAction>();
-                    allFunctions = new HashSet<IEdmFunction>();
-                }
-                else
-                {
-                    allNavigationProperties = new HashSet<IEdmNavigationProperty>();
-                    allActions = new HashSet<IEdmAction>();
-                    allFunctions = new HashSet<IEdmFunction>();
-                }
+                EdmStructuralTypeInfo structuralTypeInfo = new EdmStructuralTypeInfo(model, structuredType);
 
                 if (selectExpandClause == null)
                 {
-                    SelectedStructuralProperties = allStructuralProperties;
-                    SelectedComplexProperties = allComplexStructuralProperties;
-                    SelectedNavigationProperties = allNavigationProperties;
-                    SelectedActions = allActions;
-                    SelectedFunctions = allFunctions;
                     SelectAllDynamicProperties = true;
+
+                    // includes primitive, enum, complex or collection of them
```

Comment:
???  Doesn't structuralTypeInfo.AllNavigationProperties *only* include navigation properties?

Reply:
forgot to change the comments in iterations. 

---
In reply to: [355521785](https://github.com/OData/WebApi/pull/1941#discussion_r355521785) [](ancestors = 355521785)